In [21]:
import warnings
from causalnex.structure import StructureModel
warnings.filterwarnings("ignore")  # silence warnings
from IPython.lib.display import IFrame
from IPython.display import Image

from causalnex.plots import plot_structure, NODE_STYLE, EDGE_STYLE

In [3]:
sm = StructureModel()

In [4]:
sm.add_edges_from([
    ('health', 'absences'),
    ('health', 'G1')
])

In [17]:
sm.edges

OutEdgeView([('health', 'absences'), ('health', 'G1')])

In [18]:
#make dir supporting_files if not exists 
import os 
if not os.path.exists('supporting_files'):
    os.mkdir('supporting_files')

viz = plot_structure(
    sm,
    all_node_attributes=NODE_STYLE.WEAK,
    all_edge_attributes=EDGE_STYLE.WEAK,
)


In [24]:
import pandas as pd

data = pd.read_csv('student-por.csv', delimiter=';')
data.head(5)

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0  11  11  
1      5        3      3     1     1      3        2   9  11  11  
2      4        3      2     2     3      3        6  12  13  12  
3      3        2      2     1     1      5        0  14  14  14  
4      4        3      2     1     2      5        0  11  13  13  

[5 rows x 33 columns]

In [25]:
drop_col = ['school','sex','age','Mjob', 'Fjob','reason','guardian']
data = data.drop(columns=drop_col)
data.head(5)

address famsize Pstatus  Medu  Fedu  traveltime  studytime  failures  \
0       U     GT3       A     4     4           2          2         0   
1       U     GT3       T     1     1           1          2         0   
2       U     LE3       T     1     1           1          2         0   
3       U     GT3       T     4     2           1          3         0   
4       U     GT3       T     3     3           1          2         0   

  schoolsup famsup  ... famrel freetime goout Dalc Walc health  absences  G1  \
0       yes     no  ...      4        3     4    1    1      3         4   0   
1        no    yes  ...      5        3     3    1    1      3         2   9   
2       yes     no  ...      4        3     2    2    3      3         6  12   
3        no    yes  ...      3        2     2    1    1      5         0  14   
4        no    yes  ...      4        3     2    1    2      5         0  11   

   G2  G3  
0  11  11  
1  11  11  
2  13  12  
3  14  14  
4  13  13  

[5 rows x 26 columns]

In [26]:
import numpy as np

struct_data = data.copy()
non_numeric_columns = list(struct_data.select_dtypes(exclude=[np.number]).columns)

print(non_numeric_columns)

['address', 'famsize', 'Pstatus', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic']


In [27]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

for col in non_numeric_columns:
    struct_data[col] = le.fit_transform(struct_data[col])

struct_data.head(5)

address  famsize  Pstatus  Medu  Fedu  traveltime  studytime  failures  \
0        1        0        0     4     4           2          2         0   
1        1        0        1     1     1           1          2         0   
2        1        1        1     1     1           1          2         0   
3        1        0        1     4     2           1          3         0   
4        1        0        1     3     3           1          2         0   

   schoolsup  famsup  ...  famrel  freetime  goout  Dalc  Walc  health  \
0          1       0  ...       4         3      4     1     1       3   
1          0       1  ...       5         3      3     1     1       3   
2          1       0  ...       4         3      2     2     3       3   
3          0       1  ...       3         2      2     1     1       5   
4          0       1  ...       4         3      2     1     2       5   

   absences  G1  G2  G3  
0         4   0  11  11  
1         2   9  11  11  
2         6  12  13  12  
3         0  14  14  14  
4         0  11  13  13  

[5 rows x 26 columns]

In [34]:
from causalnex.structure.notears import from_pandas
sm = from_pandas(struct_data)

In [36]:
sm.edges, len(sm.edges)

(OutEdgeView([('address', 'famsize'), ('address', 'Pstatus'), ('address', 'Medu'), ('address', 'Fedu'), ('address', 'traveltime'), ('address', 'studytime'), ('address', 'failures'), ('address', 'schoolsup'), ('address', 'famsup'), ('address', 'paid'), ('address', 'activities'), ('address', 'nursery'), ('address', 'higher'), ('address', 'internet'), ('address', 'romantic'), ('address', 'famrel'), ('address', 'freetime'), ('address', 'goout'), ('address', 'Dalc'), ('address', 'Walc'), ('address', 'health'), ('address', 'absences'), ('address', 'G1'), ('address', 'G2'), ('address', 'G3'), ('famsize', 'address'), ('famsize', 'Pstatus'), ('famsize', 'Medu'), ('famsize', 'Fedu'), ('famsize', 'traveltime'), ('famsize', 'studytime'), ('famsize', 'failures'), ('famsize', 'schoolsup'), ('famsize', 'famsup'), ('famsize', 'paid'), ('famsize', 'activities'), ('famsize', 'nursery'), ('famsize', 'higher'), ('famsize', 'internet'), ('famsize', 'romantic'), ('famsize', 'famrel'), ('famsize', 'freetime'

In [31]:
viz = plot_structure(
    sm,
    all_node_attributes=NODE_STYLE.WEAK,
    all_edge_attributes=EDGE_STYLE.WEAK,
)


In [38]:
sm.remove_edges_below_threshold(0.8)

In [39]:
sm.edges, len(sm.edges)

(OutEdgeView([('address', 'absences'), ('address', 'G1'), ('Pstatus', 'famrel'), ('Pstatus', 'absences'), ('Pstatus', 'G1'), ('studytime', 'G1'), ('failures', 'absences'), ('paid', 'absences'), ('higher', 'Medu'), ('higher', 'G1'), ('internet', 'absences'), ('Dalc', 'Walc'), ('G1', 'G2'), ('G2', 'G3')]),
 14)

In [40]:
sm = from_pandas(struct_data, tabu_edges=[("higher", "Medu")], w_threshold=0.8)

In [41]:
sm.edges, len(sm.edges)

(OutEdgeView([('address', 'absences'), ('address', 'G1'), ('Pstatus', 'famrel'), ('Pstatus', 'absences'), ('Pstatus', 'G1'), ('studytime', 'G1'), ('failures', 'absences'), ('schoolsup', 'G1'), ('paid', 'absences'), ('higher', 'famrel'), ('higher', 'G1'), ('internet', 'absences'), ('Dalc', 'Walc'), ('G1', 'G2'), ('G2', 'G3')]),
 15)

In [45]:
sm.add_edge("failures", "G1")


In [46]:
sm = sm.get_largest_subgraph()

In [48]:
import networkx as nx

nx.drawing.nx_pydot.write_dot(sm, 'graph.dot')

In [50]:
from causalnex.network import BayesianNetwork
bn = BayesianNetwork(sm)

Bayesian Networks in CausalNex support only discrete distributions. Any continuous features, or features with a large number of categories, should be discretised prior to fitting the Bayesian Network. Models containing variables with many possible values will typically be badly fit, and exhibit poor performance. 
For example, consider P(G2 | G1), where G1 and G2 have possible values 0 to 20. The discrete conditional probability distribution is therefore specified using 21x21 (441) possible combinations - most of which we will be unlikely to observe.

CausalNex provides a few helper methods to make discretisation easier. Let’s start by reducing the number of categories in some of the categorical features by combining similar values. We will make numeric features categorical by discretisation, and then give the buckets meaningful labels.

In [51]:
discretised_data = data.copy()

data_vals = {col: data[col].unique() for col in data.columns}

failures_map = {v: 'no-failure' if v == [0]
                else 'have-failure' for v in data_vals['failures']}
studytime_map = {v: 'short-studytime' if v in [1,2]
                 else 'long-studytime' for v in data_vals['studytime']}

In [52]:
studytime_map

{2: 'short-studytime',
 3: 'long-studytime',
 1: 'short-studytime',
 4: 'long-studytime'}

In [53]:
discretised_data["failures"] = discretised_data["failures"].map(failures_map)
discretised_data["studytime"] = discretised_data["studytime"].map(studytime_map)

To make numeric features categorical, they must first be discretised. CausalNex provides a helper class `causalnex.discretiser.Discretiser`, which supports several discretisation methods. For our data the fixed method will be applied, providing static values that define the bucket boundaries. For example, `absences` will be discretised into the buckets < 1, 1 to 9, and >=10. Each bucket will be labelled as an integer from zero.

In [54]:
from causalnex.discretiser import Discretiser

discretised_data["absences"] = Discretiser(method="fixed",
                          numeric_split_points=[1, 10]).transform(discretised_data["absences"].values)
discretised_data["G1"] = Discretiser(method="fixed",
                          numeric_split_points=[10]).transform(discretised_data["G1"].values)
discretised_data["G2"] = Discretiser(method="fixed",
                          numeric_split_points=[10]).transform(discretised_data["G2"].values)
discretised_data["G3"] = Discretiser(method="fixed",
                          numeric_split_points=[10]).transform(discretised_data["G3"].values)

In [55]:
absences_map = {0: "No-absence", 1: "Low-absence", 2: "High-absence"}

G1_map = {0: "Fail", 1: "Pass"}
G2_map = {0: "Fail", 1: "Pass"}
G3_map = {0: "Fail", 1: "Pass"}

discretised_data["absences"] = discretised_data["absences"].map(absences_map)
discretised_data["G1"] = discretised_data["G1"].map(G1_map)
discretised_data["G2"] = discretised_data["G2"].map(G2_map)
discretised_data["G3"] = discretised_data["G3"].map(G3_map)

With the learnt structure model from earlier and the discretised data, we can now fit the probability distrbution of the Bayesian Network. The first step in this is specifying all of the states that each node can take. This can be done either from data, or providing a dictionary of node values. We use the full dataset here to avoid cases where states in our test set do not exist in the training set. For real-world applications, these states may need to be provided using the dictionary method.

In [56]:
bn = bn.fit_node_states(discretised_data)

In [58]:
# Split 90% train and 10% test
from sklearn.model_selection import train_test_split

train, test = train_test_split(discretised_data, train_size=0.9, test_size=0.1, random_state=7)
bn = bn.fit_cpds(train, method="BayesianEstimator", bayes_prior="K2")


In [59]:

bn.cpds["G1"]

failures    have-failure                                                 \
higher                no                                                  
schoolsup             no                            yes                   
studytime long-studytime short-studytime long-studytime short-studytime   
G1                                                                        
Fail                0.75        0.806452            0.5            0.75   
Pass                0.25        0.193548            0.5            0.25   

failures                                                                 \
higher               yes                                                  
schoolsup             no                            yes                   
studytime long-studytime short-studytime long-studytime short-studytime   
G1                                                                        
Fail                 0.5        0.612245            0.5            0.75   
Pass                 0.5        0.387755            0.5            0.25   

failures      no-failure                                                 \
higher                no                                                  
schoolsup             no                            yes                   
studytime long-studytime short-studytime long-studytime short-studytime   
G1                                                                        
Fail                 0.5        0.612903            0.5             0.5   
Pass                 0.5        0.387097            0.5             0.5   

failures                                                                 
higher               yes                                                 
schoolsup             no                            yes                  
studytime long-studytime short-studytime long-studytime short-studytime  
G1                                                                       
Fail            0.032967         0.15016       0.111111        0.255814  
Pass            0.967033         0.84984       0.888889        0.744186

In [60]:
#example of prediction
discretised_data.loc[18, discretised_data.columns != 'G1']
predictions = bn.predict(discretised_data, "G1")
print(f"The prediction is '{predictions.loc[18, 'G1_prediction']}'")
print(f"The ground truth is '{discretised_data.loc[18, 'G1']}'")

The prediction is 'Fail'
The ground truth is 'Fail'


In [61]:

from causalnex.evaluation import classification_report

classification_report(bn, test, "G1")

{'G1_Fail': {'precision': 0.7777777777777778,
  'recall': 0.5833333333333334,
  'f1-score': 0.6666666666666666,
  'support': 12},
 'G1_Pass': {'precision': 0.9107142857142857,
  'recall': 0.9622641509433962,
  'f1-score': 0.9357798165137615,
  'support': 53},
 'accuracy': 0.8923076923076924,
 'macro avg': {'precision': 0.8442460317460317,
  'recall': 0.7727987421383649,
  'f1-score': 0.8012232415902141,
  'support': 65},
 'weighted avg': {'precision': 0.8861721611721611,
  'recall': 0.8923076923076924,
  'f1-score': 0.8860973888496825,
  'support': 65}}

In [62]:
from causalnex.evaluation import roc_auc
roc, auc = roc_auc(bn, test, "G1")
print(auc)

0.9181065088757396


# Querying Marginals

To query the model for baseline marginals that reflect the population as a whole, a query method can be used. First let’s update our model using the complete dataset, since the one we currently have was only built from training data.

In [63]:
bn = bn.fit_cpds(discretised_data, method="BayesianEstimator", bayes_prior="K2")

In [64]:
from causalnex.inference import InferenceEngine

ie = InferenceEngine(bn)
marginals = ie.query()
marginals["G1"]

{'Fail': 0.25260687281677224, 'Pass': 0.7473931271832277}

In [65]:
import numpy as np

labels, counts = np.unique(discretised_data["G1"], return_counts=True)
list(zip(labels, counts))

[('Fail', 157), ('Pass', 492)]

In [66]:
marginals_short = ie.query({"studytime": "short-studytime"})
marginals_long = ie.query({"studytime": "long-studytime"})
print("Marginal G1 | Short Studtyime", marginals_short["G1"])
print("Marginal G1 | Long Studytime", marginals_long["G1"])

Marginal G1 | Short Studtyime {'Fail': 0.2776556433482524, 'Pass': 0.7223443566517477}
Marginal G1 | Long Studytime {'Fail': 0.15504850337837614, 'Pass': 0.8449514966216239}


# Do Calculus

In [67]:
print("distribution before do", ie.query()["higher"])
ie.do_intervention("higher",
                   {'yes': 1.0,
                    'no': 0.0})
print("distribution after do", ie.query()["higher"])

distribution before do {'no': 0.10752688172043011, 'yes': 0.8924731182795698}
distribution after do {'no': 0.0, 'yes': 0.9999999999999998}


In [68]:
ie.reset_do("higher")

In [69]:
print("marginal G1", ie.query()["G1"])
ie.do_intervention("higher",
                   {'yes': 1.0,
                    'no': 0.0})
print("updated marginal G1", ie.query()["G1"])

marginal G1 {'Fail': 0.25260687281677224, 'Pass': 0.7473931271832277}
updated marginal G1 {'Fail': 0.20682952942551894, 'Pass': 0.7931704705744809}


In this case, we can see that if 100% of students wanted to do higher education (as opposed to 90% in our data population), then we estimate that pass rate would increase from 74.7% to 79.3%.